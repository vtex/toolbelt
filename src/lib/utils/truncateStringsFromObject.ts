export function truncateStringsFromObject(element: any, maxStrSize: number, depth = 3, seenObjects: any[] = []) {
  if (element == null) {
    return element
  }

  const objType = typeof element

  if (objType === 'object') {
    if (seenObjects.includes(element)) {
      return '[circular]'
    }

    if (depth === 0) {
      return Array.isArray(element) ? `[array]` : `[object]`
    }

    seenObjects.push(element)
    Object.keys(element).forEach(key => {
      // seenObjects.slice creates a copy of seenObjects at the current state
      element[key] = truncateStringsFromObject(element[key], maxStrSize, depth - 1, seenObjects.slice())
    })

    return element
  }

  if (objType === 'string') {
    return element.length <= maxStrSize ? element : `${element.substr(0, maxStrSize)}[...TRUNCATED]`
  }

  if (objType === 'number' || objType === 'bigint' || objType === 'boolean' || objType === 'symbol') {
    return element
  }

  if (objType === 'function') {
    return `[function: ${element.name || 'anonymous'}]`
  }

  return `[${objType}]`
}
