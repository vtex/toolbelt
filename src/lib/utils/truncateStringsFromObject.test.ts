import { truncateStringsFromObject } from './truncateStringsFromObject'

it.each([
  [5, ['123456', '12345', '1234']],
  [5, { a: '123456', b: '12345', c: ['123456', { a: '123456' }] }],
  [10, undefined],
  [10, null],
  [10, { a: undefined, b: null }],
  [
    5,
    {
      a: [{ aa: '123456' }, { ab: '1234' }, { ac: '12345' }],
      b: { ba: undefined, bb: null, bc: '123456', bd: { bca: '123456', bcb: '1234567' } },
    },
  ],
  [5, { a: [true, false, undefined, null] }],
  [5, { namedFn: function fn() {} }],
  [5, { arrowFunctionNamed: () => {} }],
  [5, [() => {}]],
  [5, { a: Symbol('symbol description'), b: Symbol('symbol description 2') }],
])('Works on testcase %# - String limit %d', (limit: number, obj: any) => {
  expect(truncateStringsFromObject(obj, limit)).toMatchSnapshot()
})

describe('Circular checkings', () => {
  it.each([
    [
      () => {
        const obj: any = { a: '123456', b: {} }
        obj.b.c = obj
        return obj
      },
      { a: '12345[...TRUNCATED]', b: { c: '[circular]' } },
    ],
    [
      () => {
        const obj: any = { a: '123456', b: { c: {}, d: 'abcdef' } }
        obj.b.c = obj.b
        return obj
      },
      { a: '12345[...TRUNCATED]', b: { c: '[circular]', d: 'abcde[...TRUNCATED]' } },
    ],
  ] as [() => any, any][])(`Doesn't throw on circular %#`, (objCreator, expected) => {
    expect(truncateStringsFromObject(objCreator(), 5)).toStrictEqual(expected)
  })

  it.each([
    [
      () => {
        const obj: any = { a: 'abc', b: { d: '123456' }, c: {} }
        obj.c = obj.b
        return obj
      },
      { a: 'abc', b: { d: '12345[...TRUNCATED]' }, c: { d: '12345[...TRUNCATED]' } },
    ],
  ] as [() => any, any][])(
    `Doesn't add '[circular]' when references another node but it's not circular %#`,
    (objCreator, expected) => {
      expect(truncateStringsFromObject(objCreator(), 5)).toStrictEqual(expected)
    }
  )
})
