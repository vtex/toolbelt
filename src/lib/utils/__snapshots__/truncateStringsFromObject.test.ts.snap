// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Works on testcase 0 - String limit 5 1`] = `
Array [
  "12345[...TRUNCATED]",
  "12345",
  "1234",
]
`;

exports[`Works on testcase 1 - String limit 5 1`] = `
Object {
  "a": "12345[...TRUNCATED]",
  "b": "12345",
  "c": Array [
    "12345[...TRUNCATED]",
    Object {
      "a": "12345[...TRUNCATED]",
    },
  ],
}
`;

exports[`Works on testcase 2 - String limit 10 1`] = `undefined`;

exports[`Works on testcase 3 - String limit 10 1`] = `null`;

exports[`Works on testcase 4 - String limit 10 1`] = `
Object {
  "a": undefined,
  "b": null,
}
`;

exports[`Works on testcase 5 - String limit 5 1`] = `
Object {
  "a": Array [
    Object {
      "aa": "12345[...TRUNCATED]",
    },
    Object {
      "ab": "1234",
    },
    Object {
      "ac": "12345",
    },
  ],
  "b": Object {
    "ba": undefined,
    "bb": null,
    "bc": "12345[...TRUNCATED]",
    "bd": Object {
      "bca": "12345[...TRUNCATED]",
      "bcb": "12345[...TRUNCATED]",
    },
  },
}
`;

exports[`Works on testcase 6 - String limit 5 1`] = `
Object {
  "a": Array [
    true,
    false,
    undefined,
    null,
  ],
}
`;

exports[`Works on testcase 7 - String limit 5 1`] = `
Object {
  "namedFn": "[function: fn]",
}
`;

exports[`Works on testcase 8 - String limit 5 1`] = `
Object {
  "arrowFunctionNamed": "[function: arrowFunctionNamed]",
}
`;

exports[`Works on testcase 9 - String limit 5 1`] = `
Array [
  "[function: anonymous]",
]
`;

exports[`Works on testcase 10 - String limit 5 1`] = `
Object {
  "a": Symbol(symbol description),
  "b": Symbol(symbol description 2),
}
`;
